modules/ipc/include/hephaestus/ipc/zenoh/service.h

```diff
@@ -4,30 +4,200 @@
+  std::string topic;
+  ReplyT value;
+};
+
+template <typename RequestT, typename ReplyT>
+auto callService(Session& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::optional<std::chrono::milliseconds>& timeout = std::nullopt)
+    -> std::vector<ServiceResponse<ReplyT>>;
+
+// --------- Implementation ----------
+
+namespace internal {
+
+// TODO: Remove these functions once zenoh resolves the issue of changing buffers based on encoding.
+static constexpr int CHANGING_BYTES = 19;
+
+auto addChangingBytes(std::vector<std::byte> buffer) -> std::vector<std::byte>;
+auto removeChangingBytes(std::span<const std::byte> buffer) -> std::span<const std::byte>;
+
+template <typename RequestT, typename ReplyT>
+constexpr void checkTemplatedTypes() {
+  static_assert(serdes::ProtobufSerializable<RequestT> || std::is_same_v<RequestT, std::string>,
+                "Request needs to be serializable or std::string.");
+  static_assert(serdes::ProtobufSerializable<ReplyT> || std::is_same_v<ReplyT, std::string>,
+                "Reply needs to be serializable or std::string.");
+}
+
+template <class RequestT>
+auto deserializeRequest(const zenohc::Query& query) -> RequestT {
+  if constexpr (std::is_same_v<RequestT, std::string>) {
+    CHECK(query.get_value().get_encoding() == Z_ENCODING_PREFIX_TEXT_PLAIN)
+        << "Encoding for std::string should be Z_ENCODING_PREFIX_TEXT_PLAIN";
+    return static_cast<std::string>(query.get_value().as_string_view());
+  } else {
+    CHECK(query.get_value().get_encoding() == Z_ENCODING_PREFIX_EMPTY)
+        << "Encoding for binary types should be Z_ENCODING_PREFIX_EMPTY";
+    auto payload = query.get_value().get_payload();
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+    buffer = internal::removeChangingBytes(buffer);
+    DLOG(INFO) << "Deserializing buffer of size: " << buffer.size();
+    DLOG(INFO) << fmt::format("Payload: {}", fmt::join(buffer, ","));
+    RequestT request;
+    serdes::deserialize<RequestT>(buffer, request);
+    return request;
+  }
+}
+
+template <class ReplyT>
+auto onReply(zenohc::Reply&& reply, std::vector<ServiceResponse<ReplyT>>& reply_messages, std::mutex& m,
+             const std::string& topic_name) -> void {
+  const auto result = std::move(reply).get();
+  if (const auto& error = std::get_if<zenohc::ErrorMessage>(&result)) {
+    LOG(ERROR) << fmt::format("Error on service reply for '{}': '{}'", topic_name, error->as_string_view());
+    return;
+  }
+
+  const auto& sample = std::get_if<zenohc::Sample>(&result);
+  DLOG(INFO) << fmt::format("Received answer of '{}'", sample->get_keyexpr().as_string_view());
+  if constexpr (std::is_same_v<ReplyT, std::string>) {
+    CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_TEXT_PLAIN)
+        << "Encoding for std::string should be Z_ENCODING_PREFIX_TEXT_PLAIN";
```

Comment:
should we throw an exception maybe? I feel we need to be consistent in error handling